### PR TITLE
feat: watermark otwatcher enhancement

### LINKED ISSUE #345: Reduce JetStream connections/consumers used by watermark fetcher
# Summary

Today, during watermark propagation, each replica of Vertex N starts a watermark fetcher to watch OT data from Vertex (N-1), depends on how many replicas Vertex (N-1) has, the watermark fetcher runs same number of JetStream K/V watchers, thus same numbers of JetStream connections and consumers. For example, Vertex N has 10 Pods, vertex (N-1) has 5 Pods, then the consumers to OT bucket of V (N-1) is `10 * 5 = 50`. For a Source Vertex OT bucket, the consumer number is always square of replica number. If it's a Kafka source, then it will be square of partitions - a `36` partition Kafka topic source will have `1296` consumers. A `100` replica source vertex will have `10, 000` consumers, which might overwhelm the JetStream server.

We need to enhance the watermark fetcher logic to reduce the connections/consumers to JetStream.

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by chromevoid**
Fixes #345 

# Test
```shell
~ # js s report #in(1) cat(1) out(1)
Obtaining Stream stats

╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                              Stream Report                                                                                               │
├─────────────────────────────────────────────────────────────────┬─────────┬───────────┬───────────┬──────────┬─────────┬──────┬─────────┬────────────────────────────────────────────────────────────────┤
│ Stream                                                          │ Storage │ Placement │ Consumers │ Messages │ Bytes   │ Lost │ Deleted │ Replicas                                                       │
├─────────────────────────────────────────────────────────────────┼─────────┼───────────┼───────────┼──────────┼─────────┼──────┼─────────┼────────────────────────────────────────────────────────────────┤
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_PROCESSORS  │ File    │           │ 1         │ 1        │ 125 B   │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_OT          │ File    │           │ 1         │ 5        │ 615 B   │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_PROCESSORS   │ File    │           │ 2         │ 5        │ 720 B   │ 0    │ 34      │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_PROCESSORS    │ File    │           │ 2         │ 10       │ 1.4 KiB │ 0    │ 4       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_PROCESSORS │ File    │           │ 1         │ 10       │ 1.4 KiB │ 0    │ 4       │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_OT           │ File    │           │ 2         │ 25       │ 3.0 KiB │ 0    │ 89      │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_OT            │ File    │           │ 2         │ 50       │ 5.9 KiB │ 0    │ 13      │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_OT         │ File    │           │ 1         │ 50       │ 6.0 KiB │ 0    │ 20      │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ simple-pipeline-default-simple-pipeline-cat-out                 │ File    │           │ 1         │ 2,165    │ 471 KiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ simple-pipeline-default-simple-pipeline-in-cat                  │ File    │           │ 1         │ 2,165    │ 499 KiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
╰─────────────────────────────────────────────────────────────────┴─────────┴───────────┴───────────┴──────────┴─────────┴──────┴─────────┴────────────────────────────────────────────────────────────────╯

~ # js s report #in(3) cat(1) out(1)
Obtaining Stream stats

╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                              Stream Report                                                                                               │
├─────────────────────────────────────────────────────────────────┬─────────┬───────────┬───────────┬──────────┬─────────┬──────┬─────────┬────────────────────────────────────────────────────────────────┤
│ Stream                                                          │ Storage │ Placement │ Consumers │ Messages │ Bytes   │ Lost │ Deleted │ Replicas                                                       │
├─────────────────────────────────────────────────────────────────┼─────────┼───────────┼───────────┼──────────┼─────────┼──────┼─────────┼────────────────────────────────────────────────────────────────┤
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_PROCESSORS  │ File    │           │ 1         │ 1        │ 125 B   │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_OT          │ File    │           │ 1         │ 5        │ 615 B   │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_PROCESSORS   │ File    │           │ 2         │ 5        │ 720 B   │ 0    │ 129     │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_PROCESSORS    │ File    │           │ 2         │ 10       │ 1.4 KiB │ 0    │ 98      │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_PROCESSORS │ File    │           │ 3         │ 10       │ 1.4 KiB │ 0    │ 98      │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_OT           │ File    │           │ 2         │ 25       │ 3.0 KiB │ 0    │ 473     │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_OT            │ File    │           │ 2         │ 50       │ 5.9 KiB │ 0    │ 403     │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_OT         │ File    │           │ 3         │ 50       │ 6.0 KiB │ 0    │ 495     │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ simple-pipeline-default-simple-pipeline-cat-out                 │ File    │           │ 1         │ 4,560    │ 995 KiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ simple-pipeline-default-simple-pipeline-in-cat                  │ File    │           │ 1         │ 4,560    │ 1.0 MiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
╰─────────────────────────────────────────────────────────────────┴─────────┴───────────┴───────────┴──────────┴─────────┴──────┴─────────┴────────────────────────────────────────────────────────────────╯

~ # js s report #in(3) cat(4) out(1)
Obtaining Stream stats

╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                              Stream Report                                                                                               │
├─────────────────────────────────────────────────────────────────┬─────────┬───────────┬───────────┬──────────┬─────────┬──────┬─────────┬────────────────────────────────────────────────────────────────┤
│ Stream                                                          │ Storage │ Placement │ Consumers │ Messages │ Bytes   │ Lost │ Deleted │ Replicas                                                       │
├─────────────────────────────────────────────────────────────────┼─────────┼───────────┼───────────┼──────────┼─────────┼──────┼─────────┼────────────────────────────────────────────────────────────────┤
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_PROCESSORS  │ File    │           │ 1         │ 1        │ 125 B   │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_OT          │ File    │           │ 1         │ 5        │ 615 B   │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_PROCESSORS   │ File    │           │ 2         │ 5        │ 720 B   │ 0    │ 132     │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_PROCESSORS    │ File    │           │ 5         │ 10       │ 1.4 KiB │ 0    │ 110     │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_PROCESSORS │ File    │           │ 3         │ 10       │ 1.4 KiB │ 0    │ 110     │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_OT           │ File    │           │ 2         │ 25       │ 3.0 KiB │ 0    │ 497     │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_OT            │ File    │           │ 5         │ 50       │ 5.9 KiB │ 0    │ 455     │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_OT         │ File    │           │ 3         │ 50       │ 6.0 KiB │ 0    │ 549     │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ simple-pipeline-default-simple-pipeline-cat-out                 │ File    │           │ 1         │ 4,830    │ 1.0 MiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ simple-pipeline-default-simple-pipeline-in-cat                  │ File    │           │ 1         │ 4,830    │ 1.1 MiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
╰─────────────────────────────────────────────────────────────────┴─────────┴───────────┴───────────┴──────────┴─────────┴──────┴─────────┴────────────────────────────────────────────────────────────────╯

~ # js s report #in(3) cat(4) out(2)
Obtaining Stream stats

╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                              Stream Report                                                                                               │
├─────────────────────────────────────────────────────────────────┬─────────┬───────────┬───────────┬──────────┬─────────┬──────┬─────────┬────────────────────────────────────────────────────────────────┤
│ Stream                                                          │ Storage │ Placement │ Consumers │ Messages │ Bytes   │ Lost │ Deleted │ Replicas                                                       │
├─────────────────────────────────────────────────────────────────┼─────────┼───────────┼───────────┼──────────┼─────────┼──────┼─────────┼────────────────────────────────────────────────────────────────┤
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_PROCESSORS  │ File    │           │ 1         │ 2        │ 250 B   │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_PROCESSORS   │ File    │           │ 3         │ 5        │ 645 B   │ 0    │ 179     │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_OT          │ File    │           │ 1         │ 10       │ 1.2 KiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_PROCESSORS    │ File    │           │ 5         │ 10       │ 1.4 KiB │ 0    │ 145     │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_PROCESSORS │ File    │           │ 3         │ 10       │ 1.4 KiB │ 0    │ 145     │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_OT           │ File    │           │ 3         │ 25       │ 3.0 KiB │ 0    │ 729     │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_OT            │ File    │           │ 5         │ 50       │ 5.9 KiB │ 0    │ 617     │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_OT         │ File    │           │ 3         │ 50       │ 6.0 KiB │ 0    │ 723     │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ simple-pipeline-default-simple-pipeline-cat-out                 │ File    │           │ 1         │ 5,700    │ 1.2 MiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ simple-pipeline-default-simple-pipeline-in-cat                  │ File    │           │ 1         │ 5,700    │ 1.3 MiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
╰─────────────────────────────────────────────────────────────────┴─────────┴───────────┴───────────┴──────────┴─────────┴──────┴─────────┴────────────────────────────────────────────────────────────────╯

~ # js s report #in(1) cat(1) out(1)
Obtaining Stream stats

╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                              Stream Report                                                                                               │
├─────────────────────────────────────────────────────────────────┬─────────┬───────────┬───────────┬──────────┬─────────┬──────┬─────────┬────────────────────────────────────────────────────────────────┤
│ Stream                                                          │ Storage │ Placement │ Consumers │ Messages │ Bytes   │ Lost │ Deleted │ Replicas                                                       │
├─────────────────────────────────────────────────────────────────┼─────────┼───────────┼───────────┼──────────┼─────────┼──────┼─────────┼────────────────────────────────────────────────────────────────┤
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_PROCESSORS  │ File    │           │ 1         │ 2        │ 275 B   │ 0    │ 2       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_PROCESSORS   │ File    │           │ 2         │ 5        │ 720 B   │ 0    │ 212     │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_OT          │ File    │           │ 1         │ 10       │ 1.2 KiB │ 0    │ 20      │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_PROCESSORS    │ File    │           │ 2         │ 10       │ 1.4 KiB │ 0    │ 173     │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_PROCESSORS │ File    │           │ 1         │ 10       │ 1.4 KiB │ 0    │ 173     │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_OT           │ File    │           │ 2         │ 25       │ 3.0 KiB │ 0    │ 977     │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_OT            │ File    │           │ 2         │ 50       │ 5.9 KiB │ 0    │ 746     │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_OT         │ File    │           │ 1         │ 50       │ 6.0 KiB │ 0    │ 854     │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ simple-pipeline-default-simple-pipeline-cat-out                 │ File    │           │ 1         │ 6,356    │ 1.4 MiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ simple-pipeline-default-simple-pipeline-in-cat                  │ File    │           │ 1         │ 6,360    │ 1.4 MiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
╰─────────────────────────────────────────────────────────────────┴─────────┴───────────┴───────────┴──────────┴─────────┴──────┴─────────┴────────────────────────────────────────────────────────────────╯

~ # js s report #in(0) cat(0) out(0)
Obtaining Stream stats

╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                              Stream Report                                                                                               │
├─────────────────────────────────────────────────────────────────┬─────────┬───────────┬───────────┬──────────┬─────────┬──────┬─────────┬────────────────────────────────────────────────────────────────┤
│ Stream                                                          │ Storage │ Placement │ Consumers │ Messages │ Bytes   │ Lost │ Deleted │ Replicas                                                       │
├─────────────────────────────────────────────────────────────────┼─────────┼───────────┼───────────┼──────────┼─────────┼──────┼─────────┼────────────────────────────────────────────────────────────────┤
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_PROCESSORS  │ File    │           │ 1         │ 2        │ 300 B   │ 0    │ 7       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_PROCESSORS   │ File    │           │ 1         │ 5        │ 745 B   │ 0    │ 216     │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_OT          │ File    │           │ 1         │ 10       │ 1.2 KiB │ 0    │ 28      │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_PROCESSORS    │ File    │           │ 1         │ 10       │ 1.4 KiB │ 0    │ 176     │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_PROCESSORS │ File    │           │ 0         │ 10       │ 1.5 KiB │ 0    │ 176     │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_OT           │ File    │           │ 1         │ 25       │ 3.0 KiB │ 0    │ 985     │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_OT            │ File    │           │ 1         │ 50       │ 5.9 KiB │ 0    │ 754     │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_OT         │ File    │           │ 0         │ 50       │ 6.0 KiB │ 0    │ 866     │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ simple-pipeline-default-simple-pipeline-cat-out                 │ File    │           │ 1         │ 6,416    │ 1.4 MiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ simple-pipeline-default-simple-pipeline-in-cat                  │ File    │           │ 1         │ 6,420    │ 1.4 MiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
╰─────────────────────────────────────────────────────────────────┴─────────┴───────────┴───────────┴──────────┴─────────┴──────┴─────────┴────────────────────────────────────────────────────────────────╯

~ # js s report #in(1) cat(1) out(1)
Obtaining Stream stats

╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                              Stream Report                                                                                               │
├─────────────────────────────────────────────────────────────────┬─────────┬───────────┬───────────┬──────────┬─────────┬──────┬─────────┬────────────────────────────────────────────────────────────────┤
│ Stream                                                          │ Storage │ Placement │ Consumers │ Messages │ Bytes   │ Lost │ Deleted │ Replicas                                                       │
├─────────────────────────────────────────────────────────────────┼─────────┼───────────┼───────────┼──────────┼─────────┼──────┼─────────┼────────────────────────────────────────────────────────────────┤
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_PROCESSORS  │ File    │           │ 1         │ 2        │ 275 B   │ 0    │ 13      │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_PROCESSORS   │ File    │           │ 2         │ 5        │ 720 B   │ 0    │ 221     │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-out_SINK_OT          │ File    │           │ 1         │ 10       │ 1.2 KiB │ 0    │ 40      │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_PROCESSORS    │ File    │           │ 2         │ 10       │ 1.4 KiB │ 0    │ 180     │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_PROCESSORS │ File    │           │ 1         │ 10       │ 1.4 KiB │ 0    │ 180     │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ KV_simple-pipeline-default-simple-pipeline-cat-out_OT           │ File    │           │ 2         │ 25       │ 3.0 KiB │ 0    │ 997     │ isbsvc-default-js-0*, isbsvc-default-js-1, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in-cat_OT            │ File    │           │ 2         │ 50       │ 5.9 KiB │ 0    │ 767     │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ KV_simple-pipeline-default-simple-pipeline-in_SOURCE_OT         │ File    │           │ 1         │ 50       │ 6.0 KiB │ 0    │ 886     │ isbsvc-default-js-0, isbsvc-default-js-1, isbsvc-default-js-2* │
│ simple-pipeline-default-simple-pipeline-cat-out                 │ File    │           │ 1         │ 6,525    │ 1.4 MiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
│ simple-pipeline-default-simple-pipeline-in-cat                  │ File    │           │ 1         │ 6,525    │ 1.5 MiB │ 0    │ 0       │ isbsvc-default-js-0, isbsvc-default-js-1*, isbsvc-default-js-2 │
╰─────────────────────────────────────────────────────────────────┴─────────┴───────────┴───────────┴──────────┴─────────┴──────┴─────────┴────────────────────────────────────────────────────────────────╯

~ # 
```

### CODE REVIEW

File: pkg/watermark/fetch/processor_manager.go [hunk 1]
```diff
@@ -212,3 +212,52 @@ func (v *ProcessorManager) startHeatBeatWatcher() {
...
+			switch value.Operation() {
+			case store.KVPut:
+				retries := 1
+				p := v.GetProcessor(value.Key())
+				for p == nil && retries <= 10 {
+					// it is possible that by the time we receive the ot KVPut event,
+					// the heartbeat store hasn't added the new processor yet,
+					// so we need to set up the retry logic:
+					//   default processor heartbeat rate is 5 seconds
+					//   (1+..+10)*100=5500ms=5.5s
```
Comment (chromevoid) [inline pkg/watermark/fetch/processor_manager.go:236]: open to better solution

Comment (vigith) [reply pkg/watermark/fetch/processor_manager.go:236]: when a pod starts up, it first sends the heartbeat event before sending the OT event, right?

File: pkg/watermark/fetch/processor_manager.go [hunk 2]
```diff
@@ -212,3 +212,52 @@ func (v *ProcessorManager) startHeatBeatWatcher() {
...
+			return
+		case value := <-watchCh:
+			if value == nil {
+				continue
+			}
+			switch value.Operation() {
+			case store.KVPut:
+				retries := 1
+				p := v.GetProcessor(value.Key())
+				for p == nil && retries <= 10 {
```
Comment (vigith) [inline pkg/watermark/fetch/processor_manager.go:231]: what if we log as an Error and skip? did you see a case where you saw that the processor was not updated but got an OT event?

File: pkg/watermark/fetch/processor_manager.go (hunk 1 above)
Comment (chromevoid) [reply pkg/watermark/fetch/processor_manager.go:236]: Yes. There’s a small(?) chance that an OT put kv event happens when we haven’t completed processing heartbeat put kv events.

**Review by vigith — COMMENTED** [review]
Awesome 😄

File: pkg/watermark/fetch/processor_manager.go [hunk 3]
```diff
@@ -69,6 +70,7 @@ func NewProcessorManager(ctx context.Context, watermarkStoreWatcher store.Waterm
 	}
 	go v.startRefreshingProcessors()
 	go v.startHeatBeatWatcher()
+	go v.startTimeLineWatcher()
```
[the remark this answers is not in the document]
Comment (chromevoid) [reply pkg/watermark/fetch/processor_manager.go:73]: Same as before: when the V(n) pod dies (gracefully), it closes the hbwatcher and the otwatcher.

A dead V(n-1) pod won’t trigger the close watcher anymore as we use one watcher for all V(n-1) pods.

File: pkg/watermark/fetch/processor_manager.go (hunk 2 above)
Comment (chromevoid) [reply pkg/watermark/fetch/processor_manager.go:231]: > did you see a case where you saw that the processor was not updated but got an OT event?

So far only in my unit tests.

File: pkg/watermark/fetch/processor_manager.go (hunk 1 above)
Comment (chromevoid) [reply pkg/watermark/fetch/processor_manager.go:236]: Hmm if that pod continues putting new ot events then after some time it will eventually show up. But the first few events will be lost.

File: pkg/watermark/fetch/processor_manager.go (hunk 2 above)
Comment (vigith) [reply pkg/watermark/fetch/processor_manager.go:231]: let's use `ExponentialBackoff` with a cap? it will look cleaner.

File: pkg/watermark/fetch/processor_to_fetch.go [hunk 4]
```diff
@@ -74,10 +72,8 @@ func NewProcessorToFetch(ctx context.Context, processor processor.ProcessorEntit
 		entity:         processor,
 		status:         _active,
 		offsetTimeline: NewOffsetTimeline(ctx, capacity),
-		otWatcher:      watcher,
```
Comment (whynowy) [inline pkg/watermark/fetch/processor_to_fetch.go:77]: Line 70, 77 can be removed, right?

**Comment by whynowy** [comment]
Some questions:

1. In heartbeat watcher, shall we skip those ones are already expired? - imagine a vertex had 3 pods originally, then the pipeline was paused or down for some reason, all the pods went away, later on it started with 1 pod, but all the history OT would also come in.
2. When we see a OT before heartbeat comes, can we `addProcessor`? What about it's a history OT?

**Comment by chromevoid** [comment]
> Some questions:
> 
> 1. In heartbeat watcher, shall we skip those ones are already expired? - imagine a vertex had 3 pods originally, then the pipeline was paused or down for some reason, all the pods went away, later on it started with 1 pod, but all the history OT would also come in.
> 2. When we see a OT before heartbeat comes, can we `addProcessor`? What about it's a history OT?

1. Discussed offline

2. 
If we see an `OT KVPut` from a V(n-1) pod `p1` before it's `heartbeat KVPut` comes..
case 1: If this is a brand new pod, `heartbeat KVPut` happens always before `OT KVPut`. `ProcessorManager.processors map[string]*ProcessorToFetch` never has this pod before. If `heartbeat KVPut` hasn't completed processing `add new processor` by the time we see `OT KVPut`, then we have the retry logic to wait a little bit for the `heartbeat KVPut` to complete `add new processor`.

case 2: If this is a new pod, but the pod name is the same as a previous deleted pod.
This new pod will have a history OT. But does it matter? since OT is monotonically increasing..

Actually there's one more corner case: If `p1` has been deleted in the k8s namespace and is removed from V(n) `ProcessorManager.heartbeat`, but the offset timeline is still preserved, the pod reference still exists => this new pod will replace the previous pod reference, and _the old offset timeline is lost. (this could be a problem?..)_ 
example: say the old offset timeline of a deleted pod is `[1651161600300:102] -> [1651161600200:101] -> [1651161600000:100]`, and the offset timeline hasn't been all consumed yet

a new pod with the same pod name comes in and replaces the old pod reference, the old offset timeline is gone, 
and then a history(or new) OT `[1651161600500:103]` comes in, then the offset timeline for the new processor will be `[1651161600500:103] -> [-1:-1] -> [-1:-1]`... 
In short: `[1651161600300:102] -> [1651161600200:101] -> [1651161600000:100]` is lost.. (this could be a problem? as the watermark for a previous ts will be -1 based on the new pod...)

**Comment by vigith** [comment]
> In short: `[1651161600300:102] -> [1651161600200:101] -> [1651161600000:100]` is lost.. (this could be a problem? as the watermark for a previous ts will be -1 based on the new pod...)

This should be fine, time will heal it.

**Review by vigith — APPROVED** [review]

**Review by whynowy — APPROVED** [review]
